File: 50. Pow(x, n).py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1
        if n == 1:
            return x
        if n < 0:
            return 1/ self.myPow(x, -n)
        if n % 2:
            return x * self.myPow(x*x, n//2)
        return self.myPow(x*x, n//2)

class Solution:
    def myPow(self, x: float, n: int) -> float:
        '''
        iterative method
        '''
        if n < 0:
            x, n = 1/x, -n
        ans = 1
        cur = x
        while n > 0:
            if n % 2:
                ans *= cur
            cur *=cur
            n = n//2
        return ans

File: 50. Pow(x, n).java
/**
 * recursive fast pow
 * O(logn) time and space
 */
class Solution {
    public double myPow(double x, int n) {
        if (n<0) {
            x = 1/x;
            n = -n;
        }
        return fastPow(x, n);
    }
    private double fastPow(double x, int n) {
        if (n==0) {
            return 1.0;
        }
        double half = fastPow(x, n/2);
        if (n%2==0) {
            return half * half;
        } else {
            return half * half * x;
        }
    }
}

/**
 * Find all '1' bit in n, and multiply the ans with x^(2^pos)
 * Time O(logn)
 * Space O(1)
 */
class Solution {
    public double myPow(double x, int n) {
        if (n<0) {
            x = 1/x;
            n = -n;
        }
        double ans = 1.0;
        double curr = x;
        for (int i=n; i>0; i/=2) {
            if (i%2==1) {
                ans = ans * curr;
            }
            curr = curr * curr;
        }
        return ans;
    }
}